src/pulsar_producer.erl

```diff
@@ -49,233 +90,634 @@ callback_mode() -> [state_functions].
     {reuseaddr, true},
     {send_timeout, ?TIMEOUT}]).
 
--record(state, {partitiontopic,
-                broker_server,
-                sock,
-                request_id = 1,
-                producer_id = 1,
-                sequence_id = 1,
-                producer_name,
-                opts = #{},
-                callback,
-                batch_size = 0,
-                requests = #{},
-                last_bin = <<>>}).
+-type state() :: #{
+                   batch_size := non_neg_integer(),
+                   broker_server := {binary(), pos_integer()},
+                   callback := undefined | mfa() | fun((map()) -> ok),
+                   last_bin := binary(),
+                   opts := map(),
+                   partitiontopic := binary(),
+                   producer_id := integer(),
+                   producer_name := atom(),
+                   replayq := replayq:q(),
+                   request_id := integer(),
+                   requests := #{sequence_id() => [{replayq:ack_ref(),
+                                                    [gen_statem:from()],
```

Comment:
?
```suggestion
                                                    [gen_statem:from() | undefined | pid()],
```

Reply:
I think `pid()` is not possible in this case?

The sync and async calls are `gen_statem:from() | undefined`, and I think that the migration-compatible `call/cast` clauses also receive only `gen_statem:from() | undefined`?